sys/malloc_monitor/malloc_monitor.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2024 TU Dresden
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @brief   monitor heap memory usage (calls to malloc/realloc/free)
+ * @author  Mikolai Gütschow <mikolai.guetschow@tu-dresden.de>
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "architecture.h"
+#include "cpu.h"
+#include "malloc_monitor.h"
+#include "malloc_monitor_internal.h"
+
+#ifndef CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE
+#define CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE 100
+#endif
+
+#ifndef CONFIG_MODULE_SYS_MALLOC_MONITOR_VERBOSE
+#define CONFIG_MODULE_SYS_MALLOC_MONITOR_VERBOSE 0
+#endif
+
+static struct {
+    void *addr[CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE];
+    size_t size[CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE];
+    size_t current;
+    size_t high_watermark;
+} malloc_monitor = {
+    .addr = {NULL},
+    .current = 0,
+    .high_watermark = 0,
+};
+
+void malloc_monitor_add(void *ptr, size_t size, uinttxtptr_t pc, char *func_prefix)
+{
+    if (ptr == NULL) return;
+    for (uint8_t i=0; i<CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE; i++) {
+        if (malloc_monitor.addr[i] == NULL) {
+            malloc_monitor.addr[i] = ptr;
+            malloc_monitor.size[i] = size;
+            malloc_monitor.current += size;
+            if (malloc_monitor.current > malloc_monitor.high_watermark)
+                malloc_monitor.high_watermark = malloc_monitor.current;
+#if CONFIG_MODULE_SYS_MALLOC_MONITOR_VERBOSE
+            printf("%salloc(%" PRIuSIZE ") @ 0x%" PRIxTXTPTR " returned %p\n",
+                func_prefix, size, pc, ptr);
+#endif
```

Comment:
I would move this printf before the for loop, as the information displayed do not depend of the number of pointers that can be monitored.

Reply:
very true, will change that!

Reply:
Good point, thanks! Adapted in https://github.com/RIOT-OS/RIOT/pull/20363/commits/4f2a3c95dcb57a3d553377b62f29acbaa91076f4

Reply:
seems good to me !